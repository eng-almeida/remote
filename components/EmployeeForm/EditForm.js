import { useRouter } from 'next/router';
import Error from 'next/error'
import { useSelector, useDispatch } from 'react-redux';

import EmployeeForm from '@/components/EmployeeForm/partials/Form';
import FormTitle from '@/components/EmployeeForm/partials/FormTitle';
import Container from './partials/Container';
import { userSelectors, userUpdate } from '@/redux/slices/users';

const EditForm = () => {
  const router = useRouter();
  const dispatch = useDispatch();
  const employee = useSelector((state) => userSelectors.selectById(state, router.query.id));

  const handleSubmit = (employee) => {
    dispatch(dispatch(userUpdate({ id: employee.id, changes: employee })));
  }
  
  if (!employee) {
    return <Error statusCode={404} />
  }

  return (
    <Container>
      <FormTitle 
        title="Edit employee" 
        description="Edit the information of your employee" 
      />
      <EmployeeForm data={employee} onSubmit={handleSubmit} />
    </Container>
  );
}

export default EditForm;