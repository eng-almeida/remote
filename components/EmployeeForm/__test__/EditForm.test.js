import { screen, fireEvent } from '@testing-library/react';
import EditForm from '../EditForm';

jest.mock('uuid', () => ({
  v4: jest.fn().mockReturnValue("1"),
}));

jest.mock('next/router', () => ({
    useRouter: jest.fn().mockImplementation(() => ({
        push: jest.fn(),
        query: {
            id: '1'
        }
    }))
}));

const mockDispatch = jest.fn();


jest.mock('react-redux', () => {
    const actualReactRedux = jest.requireActual('react-redux');
    return {
      ...actualReactRedux,
      useDispatch: () => mockDispatch,
    };
});

describe('EditForm component', () => {
    it('should dispatch userUpdate on submit', () => {
        renderWithProviders(<EditForm />, {
            store: { 
				users: { 
					ids: ['1'], 
					entities: {
						1: {
							id: '1',
							name: 'John',
							birthdate: '10/01/1990',
							role: 'Professor',
							salary: '50000',
							country: 'Portugal'
						},
					}
				} 
			}
        });
        const addButton = screen.getByText('Save');
        const nameInput = screen.getByLabelText(/Name/i);

        fireEvent.change(nameInput, { target: { value: 'John Doe'}});
        fireEvent.click(addButton);

        expect(mockDispatch).toHaveBeenCalledWith({
            payload: {
                id: "1",
                changes: {
                    birthdate: "10/01/1990", 
                    country: "Portugal", 
                    id: "1", 
                    name: "John Doe", 
                    role: "Professor", 
                    salary: "50000"
                }
            }, 
            type: "users/userUpdate"
        });
    });
});