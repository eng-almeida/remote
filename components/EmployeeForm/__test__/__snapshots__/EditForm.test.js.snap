// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm component should match snapshot 1`] = `
.c16 {
  display: inline-block;
  text-align: center;
  background: transparent;
  border: 2px solid #624DE3;
  border-radius: 50px;
  box-shadow: 0px 6px 12px rgba(98,77,227,0.3);
  padding: 12px 26px;
  color: #624DE3;
  font-size: 16px;
  line-height: 19px;
  font-family: inherit;
  min-width: 142px;
}

.c17 {
  display: inline-block;
  text-align: center;
  background: #624DE3;
  border: 2px solid #624DE3;
  border-radius: 50px;
  box-shadow: 0px 6px 12px rgba(98,77,227,0.3);
  padding: 12px 26px;
  color: #ffffff;
  font-size: 16px;
  line-height: 19px;
  font-family: inherit;
  min-width: 142px;
}

.c6 {
  font-size: 13px;
  line-height: 16px;
  display: block;
  color: #778CA3;
  margin-bottom: 4px;
  -webkit-transition: color 0.2s;
  transition: color 0.2s;
}

.c11 {
  font-size: 13px;
  line-height: 16px;
  color: #00234B;
  opacity: 0.5;
  margin-top: 8px;
}

.c10 {
  position: absolute;
  bottom: 0;
  width: 0;
  border-bottom: 1px solid #624DE3;
}

.c7 {
  position: relative;
}

.c8 {
  font-size: 20px;
  line-height: 24px;
  font-family: inherit;
  color: #00234B;
  outline: none;
  background: transparent;
  border: none;
  border-bottom: 1px solid #CED5DB;
  padding-bottom: 8px;
  width: 100%;
}

.c8::-webkit-input-placeholder,
.c8::-webkit-input-placeholder,
.c8:-ms-input-placeholder {
  color: #525F7F;
}

.c8::-moz-placeholder,
.c8::-webkit-input-placeholder,
.c8:-ms-input-placeholder {
  color: #525F7F;
}

.c8:-ms-input-placeholder,
.c8::-webkit-input-placeholder,
.c8:-ms-input-placeholder {
  color: #525F7F;
}

.c8::placeholder,
.c8::-webkit-input-placeholder,
.c8:-ms-input-placeholder {
  color: #525F7F;
}

.c8::-webkit-outer-spin-button,
.c8::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c8[type=number] {
  -moz-appearance: textfield;
}

.c8:focus + .c9 {
  -webkit-animation: cvpVum 0.2s linear;
  animation: cvpVum 0.2s linear;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c12 {
  position: relative;
}

.c13 {
  font-size: 20px;
  line-height: 24px;
  border: none;
  border-bottom: 1px solid #CED5DB;
  color: #00234B;
  outline: none;
  padding-bottom: 8px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.c13::-ms-expand {
  display: none;
}

.c14 {
  position: absolute;
  right: 16px;
  top: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 10px;
  height: 10px;
  border-left: 2px solid #CED5DB;
  border-bottom: 2px solid #CED5DB;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

.c5 {
  display: grid;
  grid-row-gap: 30px;
  margin: 48px auto 0 auto;
  width: 600px;
}

.c15 {
  background: #F9FAFF;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
  padding: 32px 0;
  margin-top: 80px;
  text-align: center;
}

.c15 > button:first-child {
  margin-right: 16px;
}

.c2 {
  display: grid;
  grid-row-gap: 4px;
  padding: 34px 36px 30px 36px;
  border-bottom: 1px solid #EAEDF3;
}

.c3 {
  color: #00234B;
  font-weight: 500;
  font-size: 22px;
  line-height: 27px;
}

.c4 {
  color: #525F7F;
  font-size: 14px;
  line-height: 22px;
}

.c0 {
  width: 1078px;
  margin: auto;
}

.c1 {
  background: #ffffff;
  border-radius: 10px;
  margin-top: 40px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <h1
        class="c3"
      >
        Edit employee
      </h1>
      <p
        class="c4"
      >
        Edit the information of your employee
      </p>
    </div>
    <form>
      <div
        class="c5"
      >
        <div>
          <label
            class="c6"
            for="name"
          >
            Name
          </label>
          <div
            class="c7"
          >
            <input
              class="c8"
              id="name"
              name="name"
              placeholder="e.g. Jane Doe"
              required=""
              value="John"
            />
            <div
              class="c9 c10"
            />
          </div>
          <div
            class="c11"
          >
            First and last name
          </div>
        </div>
        <div>
          <label
            class="c6"
            for="birthdate"
          >
            Birthdate
          </label>
          <div
            class="c7"
          >
            <input
              class="c8"
              id="birthdate"
              name="birthdate"
              pattern="\\\\d{1,2}/\\\\d{1,2}/\\\\d{4}"
              placeholder="e.g. 17/02/1990"
              required=""
              value="10/01/1990"
            />
            <div
              class="c9 c10"
            />
          </div>
          <div
            class="c11"
          >
            DD/MM/YYYY
          </div>
        </div>
        <div>
          <label
            class="c6"
            for="role"
          >
            Job Title
          </label>
          <div
            class="c7"
          >
            <input
              class="c8"
              id="role"
              name="role"
              placeholder="e.g. Product Manager"
              required=""
              value="Professor"
            />
            <div
              class="c9 c10"
            />
          </div>
          <div
            class="c11"
          >
            What is their role?
          </div>
        </div>
        <div>
          <label
            class="c6"
            for="country"
          >
            Country
          </label>
          <div
            class="c12"
          >
            <select
              class="c13"
              id="country"
              name="country"
            >
              <option
                value="Portugal"
              >
                Portugal
              </option>
              <option
                value="Spain"
              >
                Spain
              </option>
              <option
                value="France"
              >
                France
              </option>
              <option
                value="United States"
              >
                United States
              </option>
            </select>
            <span
              class="c14"
            />
          </div>
          <div
            class="c11"
          >
            Where are they base?
          </div>
        </div>
        <div>
          <label
            class="c6"
            for="salary"
          >
            Salary
          </label>
          <div
            class="c7"
          >
            <input
              class="c8"
              id="salary"
              name="salary"
              placeholder="e.g. 50000"
              required=""
              step="100"
              type="number"
              value="50000"
            />
            <div
              class="c9 c10"
            />
          </div>
          <div
            class="c11"
          >
            Gross yearly salary
          </div>
        </div>
      </div>
      <div
        class="c15"
      >
        <button
          class="c16"
        >
          Cancel
        </button>
        <button
          class="c17"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
